spec/spec_helper.rb

```diff
@@ -241,5 +241,32 @@ def initialize(*args, &block)
 
 Thread.prepend(DatadogThreadDebugger)
 
+# Enforce test time limit, to allow us to debug why some test runs get stuck in CI
+if ENV.key?('CI')
+  require 'spec/support/thread_helpers'
+
+  ThreadHelpers.with_leaky_thread_creation('Deadline thread') do
+    Thread.new do
+      sleep_time = 30 * 60 # 30 minutes
+      sleep(sleep_time)
+
+      warn "Test too longer than #{sleep_time}s to finish, aborting test run."
+      warn 'Stack trace of all running threads:'
+
+      Thread.list.select { |t| t.alive? && t != Thread.current }.each_with_index.map do |t, idx|
+        backtrace = t.backtrace
+        backtrace = '(Not available)' if backtrace.nil? || backtrace.empty?
+
+        warn "#{idx}: #{t} (#{t.class.name})",
+          'Thread Backtrace:',
+          backtrace.map { |l| "\t#{l}" }.join("\n"),
+          "\n"
```

Comment:
I half-wonder if warn will be able to write somewhere, given that it seems that one of [stdout, stderr] is getting closed.

You may want to try writing to both? And perhaps even also to a persistent file, just in case?

Reply:
😮 
```sh
# ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [x86_64-darwin20]
ruby -e 'warn "1"; STDOUT.close; warn "2"; STDERR.close; warn "3"'
1
2
3
```

![Screen Shot 2022-08-18 at 10 58 43 AM](https://user-images.githubusercontent.com/583503/185355133-7f1d2bff-ac0e-4c43-b03b-234b7ce7156a.png)

Reply:
I tried with all versions of Ruby we support and not all have this magical power:
I've updated the code to try both `warn`, which uses either STDERR or magic, and STDOUT as a fallback. This is the most reliable I think we can get here, save for writing to a file.

Reply:
Interesting. I guess maybe at some point they started `dup()`ing the file descriptor, so that even if stdout/stderr are closed, Ruby can still in an emergency write something somewhere.

Reply:
To be fair, it only really needs to work for JRuby 9.3, which is the only one I've ever seen get bitten by this issue.